internal/controller/memcached_controller.go

```diff
@@ -75,12 +76,106 @@ func (r *MemcachedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
+			log.Error(err, "5a. Update the memcached version. Unexpected error checking version")
+			return ctrl.Result{}, err
+		}
+
+		memcached.Status.CheckedVersion = version
+
+		if valid {
+			// Version is valid, use it as the initial version.
+			memcached.Status.CurrentVersion = version
+		} else {
+			log.Info("5a. Update the memcached version. Initial version is not valid", "memcached.Spec.Version", memcached.Spec.Version)
+		}
+
+		if memcached.Status.Nodes == nil {
+			memcached.Status.Nodes = []string{}
+		}
+
+		err = r.Status().Update(ctx, memcached)
+		if err != nil {
+			log.Error(err, "5a. Update the memcached version. Failed to update status")
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if version != currentVersion && version != checkedVersion {
+		// Second case: a new version was applied.
+		// Check if version in the spec is a valid version, then check if the version change is a minor upgrade, if yes write the new version to the status.
+		valid, err := isValidVersion(version)
+		if err != nil {
+			log.Error(err, "5a. Update the memcached version. Unexpected error checking version")
+			return ctrl.Result{}, err
+		}
+
+		isMinorUpgrade, err := checkMinorUpgrade(currentVersion, version)
+		if err != nil {
+			log.Error(err, "5a. Update the memcached version. Unexpected error comparing versions")
+			return ctrl.Result{}, err
+		}
+
+		if !valid || !isMinorUpgrade {
+			// New version is not valid or not a minor upgrade, write to the status that it was already checked.
+			log.Info("5a. Update the memcached version. New version is not a valid minor upgrade", "memcached.Spec.Version", version, "memcached.Status.Version", currentVersion)
+
+			memcached.Status.CheckedVersion = version
+
+			if memcached.Status.Nodes == nil {
+				memcached.Status.Nodes = []string{}
+			}
+
+			err = r.Status().Update(ctx, memcached)
+			if err != nil {
+				log.Error(err, "5a. Update the memcached version. Failed to update status")
+				return ctrl.Result{}, err
+			}
+
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		// New version is valid and is a minor upgrade. Write it as the new version.
+		log.Info("5a. Update the memcached version. New version is a valid minor upgrade, updating status", "memcached.Spec.Version", version, "memcached.Status.Version", currentVersion)
+		memcached.Status.CheckedVersion = version
+		memcached.Status.CurrentVersion = version
+
+		if memcached.Status.Nodes == nil {
+			memcached.Status.Nodes = []string{}
+		}
+
+		err = r.Status().Update(ctx, memcached)
+		if err != nil {
+			log.Error(err, "5a. Update the memcached version. Failed to update status")
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if currentVersion == "" && version == checkedVersion {
+		// Case 3: No valid version was ever applied. Prevent deployment.
+		log.Info("5a. Update the memcached version. No valid version to deploy")
+		return ctrl.Result{}, nil
+	}
```

Comment:
I'm not sure if the checkedVersion field is really needed. Wouldn't it be sufficient to determine the deployed version by looking at the deployment?

Reply:
It's true that I have a redundant field which would be "currentVersion", which mirrors what is applied in the deployment. It could be removed in favor of parsing the version from the deployment and that would simplify the reconcile loop. I was already committed to this solution when that occurred to me, and I didn't like the idea of parsing the version out of the image string, but I think the gained simplicity outweighs that argument.
The checkedVersion however serves to avoid doing the work to check if a version inputted by the user is valid (-> http request to docker hub API) if it was already checked before, so in my eyes it's justified to have it.
Looking back, I may have overcomplicated this a little bit 😄 .

Reply:
Upon further consideration it might be not be that easy to change it to get the version from the deployment without basically changing my entire approach. In my opinion it would be better to leave it as is.

Reply:
I agree this would be a larger change and is not needed atm. 

Just one thought that occurred to me regarding your comment:

> checkedVersion however serves to avoid doing the work to check if a version inputted by the user is valid (-> http request to docker hub API) 

So it functions as a cache, right? Might be an option not to externalize the cache to the status, but instead have a sort of in-memory cache inside of the controller. Sure, that won't survive restarts, but it will be good enough to rate limit your requests to the image registry.